src/traces/bar/plot.js

```diff
@@ -108,12 +125,430 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+    var textpad;
+    if(barWidth > 2 * TEXTPAD) {
+        textpad = TEXTPAD;
+        barWidth -= 2 * textpad;
+    }
+
+    // compute rotation and scale
+    var rotate,
+        scale;
+    if(textWidth <= barWidth) {
+        // no scale or rotation
+        rotate = false;
+        scale = 1;
+    }
+    else if(textHeight <= textWidth) {
+        // only scale
+        // (don't rotate to prevent having text perpendicular to the bar)
+        rotate = false;
+        scale = barWidth / textWidth;
+    }
+    else if(textHeight <= barWidth) {
+        // only rotation
+        rotate = true;
+        scale = 1;
+    }
+    else {
+        // both scale and rotation
+        // (rotation prevents having text perpendicular to the bar)
+        rotate = true;
+        scale = barWidth / textHeight;
+    }
+
+    // compute text and target positions
+    var textX = (textBB.left + textBB.right) / 2,
+        textY = (textBB.top + textBB.bottom) / 2,
+        targetWidth,
+        targetHeight,
+        targetX,
+        targetY;
+    if(rotate) {
+        targetWidth = scale * textBB.height;
+        targetHeight = scale * textBB.width;
+    }
+    else {
+        targetWidth = scale * textBB.width;
+        targetHeight = scale * textBB.height;
+    }
+
+    if(orientation === 'h') {
+        if(x1 < x0) {
+            // bar end is on the left hand side
+            targetX = x1 - textpad - targetWidth / 2;
+            targetY = (y0 + y1) / 2;
+            if(rotate) rotate = -90;  // rotate counter-clockwise
+        }
+        else {
+            targetX = x1 + textpad + targetWidth / 2;
+            targetY = (y0 + y1) / 2;
+            if(rotate) rotate = 90;  // rotate clockwise
+        }
+    }
+    else {
+        if(y1 > y0) {
+            // bar end is on the bottom
+            targetX = (x0 + x1) / 2;
+            targetY = y1 + textpad + targetHeight / 2;
+            if(rotate) rotate = -90;  // rotate counter-clockwise
+        }
+        else {
+            targetX = (x0 + x1) / 2;
+            targetY = y1 - textpad - targetHeight / 2;
+            if(rotate) rotate = 90;  // rotate clockwise
+        }
+    }
+
+    return getTransform(textX, textY, targetX, targetY, scale, rotate);
+}
+
+function getTransform(textX, textY, targetX, targetY, scale, rotate) {
+    var transformScale,
+        transformRotate,
+        transformTranslate;
+
+    if(scale < 1) transformScale = 'scale(' + scale + ') ';
+    else {
+        scale = 1;
+        transformScale = '';
+    }
+
+    transformRotate = (rotate) ?
+        'rotate(' + rotate + ' ' + textX + ' ' + textY + ') ' : '';
+
+    // Note that scaling also affects the center of the text box
+    var translateX = (targetX - scale * textX),
+        translateY = (targetY - scale * textY);
+    transformTranslate = 'translate(' + translateX + ' ' + translateY + ')';
+
+    return transformTranslate + transformScale + transformRotate;
+}
+
+function getText(trace, index) {
```

Comment:
`Drawing` is pretty messy at the moment. Though it would be nice to merge the text-drawing logic from scatter, pie and bar together at some point. No action required for this iteration.